thriftpy/parser/parser.py

```diff
@@ -880,3 +301,224 @@ def _get_ttype(inst, default_ttype=None):
     if hasattr(inst, '__dict__') and '_ttype' in inst.__dict__:
         return inst.__dict__['_ttype']
     return default_ttype
+
+
+def _make_exception(name, fields, module):
+    return _make_struct(name, fields, module, base_cls=TException)
+
+
+def _make_const(name, val, ttype):
+    return 'const', name, val, _cast(ttype)(val)
+
+
+def _add_definition(module, type, name, val, ttype):
+    module.__thrift_meta__[type + 's'].append(val)
+    module.__dict__[name] = val
+    return type, name, val, ttype
+
+
+def _add_include(module, path, loadf):
+    included = loadf(path)
+    module.__dict__[included.__name__] = included
+    module.__thrift_meta__['includes'].append(included)
+    return 'include', included
+
+
+def _lookup_symbol(module, identifier):
+    try:
+        val = module
+        for rel_name in identifier.split('.'):
+            val = getattr(val, rel_name)
+        return val
+    except AttributeError:
+        raise UnresovledReferenceError(
+            'could not resolve name {} in module {}'.format(identifier, module.__name__))
+
+
+class UnresovledReferenceError(ThriftParserError): pass
+
+
+def _ref_type(module, name):
+    'resolve a reference to a type, return the resulting ttype'
+    val = _lookup_symbol(module, name)
+    if val in BASE_TYPE_MAP.values():
+        return val
+    if isinstance(val, tuple):  # typedef
+        return val
+    return val._ttype, val  # struct or enum
+
+
+def _ref_val(module, name):
+    'resolve a reference to a value, return the value'
+    val = _lookup_symbol(module, name)
+    if isinstance(val, type):
+        raise UnresovledReferenceError("{} in {} is a type, not a value".format(name, module))
+    return val
+
+
+class NoSuchAttribute(ThriftParserError): pass
+
+
+def _attr_ttype(struct, attr):
+    'return the ttype of attr in struct'
+    if attr not in struct._tspec:
+        raise NoSuchAttribute('no attribute {} of struct {} in module {}'.format(
+            attr, struct.__name__, struct.__module__))
+    return struct._tspec[attr][1]
+
+
+GRAMMAR = '''
```

Comment:
i could see this dense of a grammar being daunting to some. I think a link to [the thrift grammar page](https://thrift.apache.org/docs/idl) and bit of whitespace could go a long way, though :)